Development

ch0ijihwan opened the pull request.

ch0ijihwan committed:
refactor : 더 직관적인 메소드 명으로 메소드 변경

ch0ijihwan committed:
refactor : 숫자야구 시작시 인풋을 새로 받는 컨트롤러 구조 변경

ch0ijihwan committed:
refactor : MVC 패던에 맞춰서, 컨트롤러 및  GameOneRound을 수정하여,
GameOneRound 모델에서 값을 바로 뷰로 보내던 문제를 해결

ch0ijihwan committed:
test : 메소드명 변경에 따른 테스트코드 수정

ch0ijihwan committed:
test : GameOneRound 클래스 생성자 변경에 따른, 테스트 코드 수정

ch0ijihwan committed:
refactor : GameOneRound클래스를 더 가독성 있게 하기 위해 Round로 클래스명 변경

ch0ijihwan committed:
refactor : 가독성을 위한 메소드 명 변경

ch0ijihwan committed:
test : 볼과 스트라이크의 개수를 제대로 반환하는지 확인하는 테스트 코드

ch0ijihwan committed:
test : 가독성을 위해 메소드 순서 변경

ch0ijihwan committed:
test : 새로운 숫자야구 넘버를 입력 받을 시 threeBalls 리스트를 생성하는지에 대한 테스트 코드

ch0ijihwan committed:
test : 소나린트에 따른 필드 위치 변경

ch0ijihwan committed:
refactor : 필드를 final로 변경

ch0ijihwan committed:
test :
생성자에서 새로운 숫자야구 넘버를 받을 때, 올바르지 않은 값을 받은경우 (입력값이 세자리가 아닌경우) 예외 처리를 하는지 확인 하는 테스트 코드 구현
및
생성자에서 새로운 숫자야구 넘버를 받을 때, 올바르지 않은 값을 받은경우 (중복된 수가 있는 경우) 예외 처리를 하는지 확인 하는 테스트 코드 구현

ch0ijihwan committed:
test : 불필요한 강제 형변환 코드 삭제

Byeongju-Kong commented on src/test/java/model/BaseballsTest.java:
```diff
@@ -40,6 +42,26 @@ void setUp() {
+    @DisplayName("생성자에서 새로운 숫자야구 넘버를 받을 때, 올바르지 않은 값을 받은경우 (입력값이 세자리가 아닌경우) 예외 처리를 하는지 확인")
+    void validateNumber(String value) {
+        String[] tokens = value.split("");
+        assertThatThrownBy(() -> baseballs = new Baseballs(tokens))
```
assertThatIllegalArgumentException() 이라는 메소드로 예외발생테스트를 할 수 있으니 참고하시면 좋을 것 같아요~!

ch0ijihwan replied:
네 알겠습니다 ^_^
그런 방식이 더 구체적인 테스트 코드를 제시하는 것 같습니다

Byeongju-Kong replied:
assertThatIllegalArgumentException()이 assertThatTrownBy().isInstanceOf 보다 좋다는 것이 아니예요~!
개인의 선호 차이인 것 같습니다.

Byeongju-Kong commented on src/main/java/model/Round.java:
```diff
@@ -3,12 +3,12 @@
     private List<BallStatus> comparedResult;
     private List<Integer> result = new ArrayList<>();
 
-    public GameOneRound(Baseballs uesrBaseballNumbers, Baseballs systemBaseballNumbers) {
```
Round가 systemBaseballNumbers를 필드로 가지고 있고 userBaseballNumbers 만을 메소드의 파라미터로 입력받아 카운트를 하는 로직은 어떨까요? 생각해보셔도 좋을 것 같습니다.

ch0ijihwan replied:
현재 로직에 적용하여 systemBaseballs를 갖고 있도록 변경 하였습니다.

ch0ijihwan committed:
refactor : 안쓰는 메소드 삭제

ch0ijihwan committed:
test : 테스트 코드 수정

불필요한 코드 줄임 및 메소드 소스를 CSV 소스로 바꿈

ch0ijihwan committed:
test : 테스트 코드 수정

테스트 값 수정,
테스트 코드의 경계값을 수정해주고, 중복 되는 테스트 부분 삭제

ch0ijihwan committed:
refactor : 오타 수정

불필요한 코드 줄임 및 메소드 소스를 CSV 소스로 바꿈

ch0ijihwan committed:
test : 오버라이드 테스트

오버라이드한 이퀄즈가 정상작동 하는지 확인 하는 테스트

ch0ijihwan committed:
test : 테스트 코드 수정

한번 밖에 동작하지 않았던, 셋업 삭제.
add로 일일히 요소를 추가해주더문제를, asList를 써서 한번에 추가해

Baseball 객체에 대한 테스트 부분 삭제

ch0ijihwan committed:
chore : 불필요한 줄 띄움 제거

ch0ijihwan committed:
test : 테스트 코드 수정

메소드 소드를 Csv소스로 변경.

테스트시 기본 baseballs와 입력받은 otherBaseballs를 비교하도록 변경

ch0ijihwan committed:
chore : 불 필요한 줄 띄움 제거

ch0ijihwan committed:
refactor : 불필요한 과정 줄임

ch0ijihwan committed:
refactor : Round 클래스 수정

Baseballs를 받아 비교를 하고, 그에 대한 값을 TotalResult로 넘겨, 볼 카운트와 스트라이크 카운트를 받아오는 구조에서

Round객체에서 바로 볼 카운트와 스트라이트 카운트를 받을 수 있도록 함.

ch0ijihwan committed:
refactor : TotalResult 클래스 제거.

ch0ijihwan committed:
refactor : 바뀐 Round 구조에 따른 코드 수정

ch0ijihwan committed:
chore : 불필요한 줄 띄움 수정

ch0ijihwan committed:
refactor : RandomBaseballNumbersGenerator 클래스 이름 수정

ch0ijihwan committed:
refactor : 랜섬 베이스볼 반환부 수정

TreeSet을 사용해 , 램덤 볼이 순서대로 나오는 문제를 수정.

ch0ijihwan committed:
refactor : 해시 코드 삭제

ch0ijihwan committed:
refactor : nothing도 반환하도록 뷰 수정

ch0ijihwan committed:
refactor : Enum 값 이름 수정

가독성을 높히기 위해 중복 되는 STATUS 제거

ch0ijihwan committed:
refactor : 메서드 이름 변경

ch0ijihwan committed:
refactor :  view에 대한 의존도를 낮춤.
String 으로 관리하던 BaseballNumber을 int형으로 관리하도록 수정,

ch0ijihwan committed:
test :
생성자로 부터 String 값이 아닌 int 형 값을 받게 함에 따른 테스트 코드 수정. 및 예외처리 범위 수정

ch0ijihwan committed:
refactor : 값을 int으로 받는 구조로 수정

ch0ijihwan committed:
chore : 불필요한 개행 삭제

ch0ijihwan committed:
refactor : 역할을 낮춤
랜덤 야구공 리스트를 반환 하는 것 아닌, 랜덤 int를 갖고 있는 int형 배열을 반환하는 구조로 변경. + 클래스 이름 변경

ch0ijihwan committed:
test : 테스트코드 수정

ch0ijihwan committed:
feat  : 숫자를 입력 받고 전달하는 DTO 생성

ch0ijihwan committed:
refactor : 주석 추가.

ch0ijihwan committed:
test : DTO에 대한 테스터 코드 생성

ch0ijihwan committed:
test : 숫자입력 스트링 타입으로 DTO로 넘겨주도록 함

ch0ijihwan committed:
refactor : 생성자 안 불필요한 행간 띄움 제거

ch0ijihwan committed:
refactor : DTO에 대한 validateNumber 제거

입력받은 문자에 대한 예외처리를 DTO에 넣어줌

ch0ijihwan committed:
refactor : 불필요한 줄 띄움 삭제

ch0ijihwan committed:
test : 불필요한 코드 제거 및 DisplayName 수정

ch0ijihwan committed:
refcator : DTO생성에 따른 컨트롤러 수정

ch0ijihwan committed:
test : DTO 생성에 따른 예외처리 부분 삭제

ch0ijihwan committed:
test : 중복되는 코드를 setUp으로 빼줌

ch0ijihwan committed:
test : DTO에 대한 테스터 코드 수정

Byeongju-Kong commented on src/main/java/model/DtoInputNumbers.java:
```diff
@@ -0,0 +1,27 @@
+    private static final int NUMBER_LENGTH = 3;
+    private int[] inputNumbers;
+    public DtoInputNumbers(String [] tokens){
+        this.inputNumbers = Arrays.stream(tokens).mapToInt(Integer::parseInt).toArray();
```
메소드를 여러개 호출 할때에는 개행을 해주면 읽기 편할 것 같아요~

ch0ijihwan replied:
변경하였습니다.

Byeongju-Kong commented on src/main/java/model/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,31 @@
+        }
+    }
+
+    public int[] getRandomNumbers() {
+        int[] randomNumbers = new int[RANDOM_NUMBER_SIZE];
+        Iterator<Integer> iterThreeNumbers = threeNumbers.iterator();
+
+        IntStream.range(0, RANDOM_NUMBER_SIZE)
+                .forEach(index -> randomNumbers[index] = iterThreeNumbers.next());
+
+        return randomNumbers;
+    }
```
iterator를 돌리는 것도 좋지만, Stream에 toArray 메소드를 사용하는 것이 더 좋을 것 같네요

ch0ijihwan replied:
변경해보았습니다~

Byeongju-Kong commented on src/main/java/model/Round.java:
```diff
@@ -1,27 +1,34 @@
-    public boolean isThreeStrike() {
-        totalResult = new TotalResult(comparedResult);
-        return totalResult.isResultThreeStrike();
+    private int countScore(BallStatus desiredValue) {
```
desire 이라는 단어 보다는 다른 단어로 대체 하는 것이 나을 것 같아요~

ch0ijihwan replied:
현재 로직에서 이 부분이 삭제되었습니다

Byeongju-Kong commented on src/main/java/model/BaseballNumber.java:
```diff
@@ -31,9 +29,4 @@ public boolean equals(Object o) {
-    @Override
-    public int hashCode() {
-        return Objects.hash(number);
-    }
```
equals는 삭제하지 않고 hashCode를 삭제한 이유가 뭔가요? 
equals의 비교연산은 hashCode를 기반으로 합니다. 그래서 equals를 오버라이드하면 hashCode가 같이 오버라이드 되는 것이고요.

Byeongju-Kong commented on src/main/java/view/StatusDisplay.java:
```diff
@@ -6,7 +6,11 @@ private StatusDisplay() {
-        System.out.println("Ball : " + ball + "   Strike : " + strike);
+        if (ball > 0 || strike > 0)
+            System.out.println("Ball : " + ball + "   Strike : " + strike);
+        if(ball ==0 && strike ==0){
```
비교연산자와 피비교연산자 사이에 띄워쓰기가 있는 것이 컨벤션입니다.

Byeongju-Kong commented on src/main/java/model/DtoInputNumbers.java:
```diff
@@ -0,0 +1,27 @@
+    private int[] inputNumbers;
+    public DtoInputNumbers(String [] tokens){
+        this.inputNumbers = Arrays.stream(tokens).mapToInt(Integer::parseInt).toArray();
+        validateNumbers(inputNumbers);
```
필드에 값을 할당하기 전에 유효검사를 먼저하는게 좋지 않을까요?

Byeongju-Kong commented on src/test/java/model/DtoInputNumbersTest.java:
```diff
@@ -0,0 +1,43 @@
+        String[] actual = tokens;
+
+        //then
+        assertThatThrownBy(() -> dtoInputNumbers = new DtoInputNumbers(actual))
```
assertThatThrownBy(() -> new DtoInputNumbers(actual))

라고만 적어둬도 될 것 같아요

ch0ijihwan replied:
현재 코드가 assertThatThrownBy(() -> new DtoInputNumbers(actual)) 형태로 되어있습니다.

Byeongju-Kong commented on src/test/java/model/DtoInputNumbersTest.java:
```diff
@@ -0,0 +1,43 @@
+
+    @ParameterizedTest
+    @CsvSource(value = {"12", "1234"})
+    @DisplayName("생성자에서 새로운 숫자야구 넘버를 받을 때, 올바르지 않은 값을 받은경우 (입력값이 세자리가 아닌경우) 예외 처리를 하는지 확인")
```
생성자에서 새로운 숫자야구 넘버를 받을 때, "입력 값이 세자리가 아닌 경우 예외 처리를 하는지 확인" 이라고 적는 것이 더 깔끔한 DisplayName이라고 생각이 듭니다. 예외를 발생시킨다는 것은 올바르지 않은 값이기에 예외를 발생시키는 것이기 때문에
 올바르지 않다( 구체적으로 올바르지 않는 이유) 와 같은 형식 보다는 구체적으로 올바르지 않은 이유만을 적는 것이 좋은 것 같네요.

 또한 Test 자체가 확인을 한다는 뜻인데, 마지막에 확인이라는 뜻을 넣지 않아도 좋을 것 같아요.

"입력값이 세자리가 아닌 숫자들로 객체를 생성하면 예외를 발생시킨다." 
와 같은 DisplayName이 훨씬 더 깔끔한 것 같습니다.

ch0ijihwan replied:
좋은 의견입니다.
조금 보태자면 '입력값이 세자리가 아닌 숫자들'보다는
입력값이 세자리가 아닌 숫자는 어떨까요?

ch0ijihwan committed:
test : Position 테스트 코드 구현
값 객체 생성

ch0ijihwan committed:
feat : Position 값 객체 생성

ch0ijihwan committed:
test : 가독성을 위한 띄어쓰기 추가

ch0ijihwan committed:
test : value() 메서드에 대한 테스트 코드 추가.

ch0ijihwan committed:
chore : 불필요한 줄 띄움 제거

ch0ijihwan committed:
refactor : 이퀄즈 오버라이드 및 필드명 수정

포지션을 비교하기 위한 equals 추가, position 필드를 파이널로 선언, 예외처리 메시지 오타 수정.

ch0ijihwan committed:
test : 테스터 코드의 가독성을 높힘

ch0ijihwan committed:
test :  테스트 코드 분리.
countBall 메서드와 countStrike 메서드를 테스트 하는 테스트 코드 분리

ch0ijihwan committed:
refactor : 소나 린트에 따른 hash코드 추가
+
오타 수정

ch0ijihwan committed:
refactor :
 checkBaseball   접근제어자 추가 및 메서드 이름 변경

 Position  값객체 추가에 따른 position에 직접 접근하는 방식을 수정

ch0ijihwan committed:
refactor : 메서드 이름 변경

ch0ijihwan committed:
test : 예외처리 메시지 수정에 따른 테스트 코드 변경

ch0ijihwan committed:
test : 테스트 코드의 가독성을 높힘

ch0ijihwan committed:
test : 테스트 코드의 가독성을 높힘

ch0ijihwan committed:
test : 필드 선언 부분 수정

ch0ijihwan committed:
chore : 불필요한 코드 수정

ch0ijihwan committed:
chore : 패키지 생성

Byeongju-Kong commented on src/main/java/model/Baseball.java:
```diff
@@ -2,20 +2,20 @@
 
-    public Baseball(BaseballNumber baseBallNumber, int position) {
-        number = baseBallNumber;
+    public Baseball(BaseballNumber baseBallNumber, Position position) {
+        this.number = baseBallNumber;
```
매개변수를 int, int가 아닌 BaseballNumber, Position으로 해준 이유가 있을까요?
Baseball 밖에서 BaseballNumber와 Position 객체를 만들어 매개변수로 넣어주기보다, 
int 타입으로 받아 Baseball에서 BaseballNumber와 Position 객체를 만들어주는 것도 방법이라고 생각합니다.
그렇게 되면 BaseballNumber와 Position의 의존성은 Baseball에서만 가지게 되겠네요.
굳이 밖에서 두 객체를 생성하면서 의존성을 줄 필요가 없다고 생각합니다.

ch0ijihwan replied:
제가 그 부분 까지 생각하기 못하였습니다.
한번 수정해보도록 하였습니다.

Byeongju-Kong replied:
불필요한 의존성을 제거하는 것이 유지보수 측면에서 중요한 것 같아요.
다른 객체들도 불필요한 의존성이 있는지 한번씩 검토해보시면 좋을 것 같습니다 :^)

Byeongju-Kong commented on src/main/java/model/Position.java:
```diff
@@ -0,0 +1,23 @@
+public class Position {
+    private static final int MIN_BOUNDARY = 0;
+    private static final int MAX_BOUNDARY = 2;
+    private int position;
```
해당 객체를 Value Object의 개념으로 생성하신 것으로 추측되는데,
그렇다면 값 객체의 값은 final이 되는 것이 훨씬 좋고 안정적이라고 생각됩니다.
물론 생성자를 제외하고 해당 필드의 값을 변경하는 일이 없는 것으로 보입니다. 그렇다고 하더라도, 또한 그렇기 때문에
final 키워드가 있으면 더 좋을 것이라는 생각이 듭니다.

ch0ijihwan replied:
일단 현재 코드는 final 이 추가된 상태입니다.

Byeongju-Kong commented on src/main/java/model/Position.java:
```diff
@@ -1,23 +1,37 @@
     private static final int MIN_BOUNDARY = 0;
     private static final int MAX_BOUNDARY = 2;
-    private int position;
+    private final int index;
```
final 키워드를 추가하셨네요.
해당 객체는 공의 위치의 개념을 가지고 있다고 생각하는데
그렇다면 position의 개념과 index의 개념은 서로 완전하게 맞지 않다고 생각이 듭니다.
이는 값 객체이기때문에 필드를 간단하게 value라고 해주어도 좋을 것 같네요

ch0ijihwan replied:
이 부분은 Postion이라는 클래스 내에  단순 값(value)라고 선언하는 것보다. 
현재 숫자야구 특성상, 공의 위치(Position)가 단순 위치가 아니라, 공이 나열된 순서를 나타내기 때문에, 필드명을 index라 하였습니다.

index하면 '순서가 있는 나열 중, 그 객체의 위치'라는 것이 떠오른다고 생각하였습니다.

Byeongju-Kong replied:
지환님의 말씀대로, index 측면에서의 순서를 나타내기 위한 객체라면 이름이 Position이 아닌 Index가 되는 것이 더 적절한 것이 아닐까요?

제가 말씀드리는 것은
Position과 index를 혼용해서 써서 개념적인 충돌이 발생할 수 있는 가능성이 조금이라도 있기 때문에,
필드의 이름을 value로 두고 Position(혹은 Index)이라는 클래스의 이름으로 필드의 개념을 알 수 있도록 하는 것이
개념적인 충돌이 발생할 확률이 적다고 생각합니다.

ch0ijihwan replied:
알겠습니다. 고려해보도록 하겠습니다.
개념적인 충돌이 일어날 수 있다는 것에 동의합니다.

ch0ijihwan committed:
test : 불필요한 과정 삭제.

ch0ijihwan committed:
test : 오타 제거

ch0ijihwan committed:
test : DisplayName 수정

ch0ijihwan committed:
test : final 추가

ch0ijihwan committed:
chore : baseball 클래스 패키지 위치 변경

Byeongju-Kong commented:
간단하게 리뷰 남겨뒀으니 확인해보세요~

Byeongju-Kong reviewed (commented):
간단하게 리뷰 남겼습니다~

ch0ijihwan committed:
refactor : 메소드 개행

ch0ijihwan committed:
refactor : iterator 방식 변경

ch0ijihwan committed:
refactor : 생성자 변경

ch0ijihwan committed:
test : DisplayName 변경

ch0ijihwan committed:
refactor : 필드명 변경
